token/services/interop/pledge/wallet.go

```diff
@@ -0,0 +1,236 @@
+		if tok.Id.String() == tokenID.String() {
+			owner, err := identity.UnmarshalTypedIdentity(tok.Owner.Raw)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "failed to unmarshal owner")
+			}
+			logger.Debugf("found token [%s] with type [%s]", tokenID, owner.Type)
+			if owner.Type == ScriptType {
+				res = append(res, tok)
+				script := &Script{}
+				if err := json.Unmarshal(owner.Identity, script); err != nil {
+					return nil, nil, errors.Wrapf(err, "failed unmarshalling pledge script")
+				}
+				scripts = append(scripts, script)
+			}
+		}
+	}
+	if len(res) > 1 {
+		return nil, nil, errors.Errorf("multiple pledged tokens with the same identifier [%s]", tokenID.String())
+	}
+	if len(res) == 0 {
+		return nil, nil, errors.Errorf("no pledged token exists with identifier [%s]", tokenID.String())
+	}
+	return res[0], scripts[0], nil
+}
+
+// ScriptAuth implements the Authorization interface for this script
+type ScriptAuth struct {
+	WalletService driver.WalletService
+}
+
+func NewScriptAuth(walletService driver.WalletService) *ScriptAuth {
+	return &ScriptAuth{WalletService: walletService}
+}
+
+// AmIAnAuditor returns false for script ownership
+func (s *ScriptAuth) AmIAnAuditor() bool {
+	return false
+}
+
+// IsMine returns true if either the sender or the recipient is in one of the owner wallets.
+// It returns an empty wallet id.
+func (s *ScriptAuth) IsMine(tok *token2.Token) (string, []string, bool) {
+	identity, err := identity.UnmarshalTypedIdentity(tok.Owner.Raw)
+	if err != nil {
+		logger.Debugf("Is Mine [%s,%s,%s]? No, failed unmarshalling [%s]", view2.Identity(tok.Owner.Raw), tok.Type, tok.Quantity, err)
+		return "", nil, false
+	}
+	if identity.Type != ScriptType {
+		return "", nil, false
+	}
+
+	script := &Script{}
+	if err := json.Unmarshal(identity.Identity, script); err != nil {
+		logger.Debugf("Is Mine [%s,%s,%s]? No, failed unmarshalling [%s]", view2.Identity(tok.Owner.Raw), tok.Type, tok.Quantity, err)
+		return "", nil, false
+	}
+	if script.Sender.IsNone() || script.Recipient.IsNone() || script.Issuer.IsNone() {
+		logger.Debugf("Is Mine [%s,%s,%s]? No, invalid content [%v]", view2.Identity(tok.Owner.Raw), tok.Type, tok.Quantity, script)
+		return "", nil, false
+	}
+
+	// I'm either a sender, recipient, or issuer
+	var ids []string
+	for _, beneficiary := range []struct {
+		identity view2.Identity
+		desc     string
+		prefix   string
+	}{
+		{
+			identity: script.Sender,
+			desc:     "sender",
+			prefix:   "pledge.sender",
+		},
+		{
+			identity: script.Recipient,
+			desc:     "recipient",
+			prefix:   "pledge.recipient",
+		},
+		{
+			identity: script.Issuer,
+			desc:     "issuer",
+			prefix:   "pledge.issuer",
+		},
+	} {
+		logger.Debugf("Is Mine [%s,%s,%s] as a %s?", view2.Identity(tok.Owner.Raw), tok.Type, tok.Quantity, beneficiary.desc)
+		// TODO: differentiate better
+		if wallet, err := s.WalletService.OwnerWallet(beneficiary.identity); err == nil {
```

Comment:
We could extend this method to take more identities in the input to avoid redundant locks in the future.